File: chapter10/lists_and_recursion_6.ex
defmodule MyList do
  def flatten(list), do: _flatten(list, [])

  defp _flatten([], acc), do: Enum.reverse(acc)
  defp _flatten([head | tail], acc) when is_list(head), do: _flatten(head, acc) ++ _flatten(tail, [])
  defp _flatten([head | tail], acc), do: _flatten(tail, [head | acc])
end

File: chapter10/lists_and_recursion_6.clj
(ns lists-and-recursion-6
  (:refer-clojure :exclude [flatten]))

(defn flatten [coll]
  (letfn [(-flatten [acc [head & tail :as coll]]
            (cond
              (empty? coll) (reverse acc)
              (sequential? head) (concat (-flatten acc head)
                                         (-flatten () tail))
              :else (-flatten (cons head acc) tail)))]
    (-flatten () coll)))
